File: src/012. Highly divisible triangular number/012.cpp
#include <iostream>

int numOfFactors(long long num);

int main() {
    long long triangle = 0;
    int i = 1;
    while (true) {
        triangle += i;
        if (numOfFactors(triangle) > 500) {
            std::cout << triangle << std::endl;
            break;
        }
        ++i;
    }
}

int numOfFactors(long long num) {
    int count = 0;
    long factor = 1;
    while (factor * factor < num) {
        if (num % factor == 0) {
            count += 1;
        }
        factor += 1;
    }
    if (factor * factor > num) {
        return count * 2;
    } else {
        // perfect square
        return count * 2 + 1;
    }
}

File: src/012. Highly divisible triangular number/012.go
package main

import "fmt"

func main() {
	triangle := 0
	i := 1
	for {
		triangle += i
		if numOfFactors(triangle) > 500 {
			fmt.Println(triangle)
			break
		}
		i++
	}
}

func numOfFactors(num int) int {
	n := 0
	factor := 1
	for factor * factor < num {
		if num % factor == 0 {
			n += 1
		}
		factor += 1
	}
	if factor * factor > num {
		return n * 2
	} else {
		// perfect square
		return n * 2 + 1
	}
}
